File: 2020/15/main_test.go
package main

import (
	"fmt"
	"testing"
)

func TestTask1ShouldFindResult(t *testing.T) {
	input := []int{0, 3, 6}
	expected := 436

	result := Task1(input)
	if result != expected {
		t.Errorf("got %d, expected %d", result, expected)
	}
}

func TestTask2ShouldFindResult(t *testing.T) {
	tests := [][]int{
		{0, 3, 6},
		{1, 3, 2},
		{2, 1, 3},
		{1, 2, 3},
		{2, 3, 1},
		{3, 2, 1},
		{3, 1, 2},
	}
	expected := []int{
		175594,
		2578,
		3544142,
		261214,
		6895259,
		18,
		362,
	}

	for i, input := range tests {
		fmt.Printf("Test %d: %+v\n", i, input)
		result := Task2(input)
		if result != expected[i] {
			t.Errorf("%d: got %d, expected %d", i, result, expected[i])
		}
	}
}

File: 2021/17/src/main_test.zig
const std = @import("std");
const builtin = @import("builtin");

const app = @import("./main.zig");

inline fn checkPrint(ok: bool) void {
    if (ok) {
        std.debug.print(" ✅\n", .{});
    } else {
        std.debug.print(" ❌\n", .{});
    }
}

test "setup" {
    if (builtin.os.tag == .windows) {
        // NOTE: Make windows use UTF-8 for output.
        _ = std.os.windows.kernel32.SetConsoleOutputCP(65001);
    }
}

test "expect task 1 to result in 45" {
    const tests = .{
        .{
            \\target area: x=20..30, y=-10..-5
            \\
            ,
            45,
        },
    };

    std.debug.print("\n", .{});
    var ok = true;
    inline for (tests) |t| {
        var stream = std.io.fixedBufferStream(t[0]);
        var input = try app.readInput(std.testing.allocator, stream.reader());

        const result = app.task1(input);
        std.debug.print("Expect: {}, got {}", .{ t[1], result });
        checkPrint(result == t[1]);
        if (result != t[1]) {
            ok = false;
        }
    }
    try std.testing.expect(ok);
}

test "expect task 2 to result in 112" {
    const tests = .{
        .{
            \\target area: x=20..30, y=-10..-5
            \\
            ,
            112,
        },
    };

    std.debug.print("\n", .{});
    var ok = true;
    inline for (tests) |t| {
        var stream = std.io.fixedBufferStream(t[0]);
        var input = try app.readInput(std.testing.allocator, stream.reader());

        const result = app.task2(input);
        std.debug.print("Expect: {}, got {}", .{ t[1], result });
        checkPrint(result == t[1]);
        if (result != t[1]) {
            ok = false;
        }
    }
    try std.testing.expect(ok);
}
